services/inc/check.h

```diff
@@ -15,6 +15,107 @@
+ *
+ * Device OS doesn't use C++ exceptions, and, by convention, its internal and public API functions
+ * return a negative value in case of an error. All error codes are defined in the system_error.h
+ * file.
+ *
+ * Consider the following example:
+ * ```
+ * int foo() {
+ *     void* buffer = malloc(1024);
+ *     if (!buffer) {
+ *         return SYSTEM_ERROR_NO_MEMORY; // -260
+ *     }
+ *     ...
+ *     free(buffer);
+ *     return 0; // No error
+ * }
+ * ```
+ * When invoking such a function, the calling function should properly handle all possible errors.
+ * In the most typical scenario, it simply forwards the error to its own calling function:
+ * ```
+ * int fooBarBaz() {
+ *     int result = foo();
+ *     if (result == 0) {
+ *         result = bar();
+ *         if (result == 0) {
+ *             result = baz();
+ *         }
+ *     }
+ *     return result;
+ * }
+ * ```
+ * When many functions are involved, writing the code like in the above example becomes tedious.
+ * Given that all the functions follow the convention, the code can be rewritten as follows using
+ * the `CHECK()` macro:
+ * ```
+ * int fooBarBaz() {
+ *     // This macro makes fooBarBaz() return SYSTEM_ERROR_NO_MEMORY if foo() fails to allocate its
+ *     // internal buffer
+ *     CHECK(foo());
+ *     CHECK(bar());
+ *     CHECK(baz());
+ *     return 0;
+ * }
+ * ```
+ * The `CHECK()` macro expands to a code that evaluates the expression passed to the macro, and if
+ * the expression yields a negative value makes the function that invokes the macro return that
+ * value. Non-negative values get passed back to the calling code:
+ * ```
+ * int parseStream(Stream* stream) {
+ *     char buffer[128];
+ *     size_t size = CHECK(stream->read(buffer, sizeof(buffer));
+ *     LOG(TRACE, "Read %u bytes", (unsigned)size);
+ *     ...
+ *     return 0;
+ * }
+ * ```
+ * In the above example, it is safe to assume that the value returned by `Stream::read()`, which is
+ * wrapped into the `CHECK()` macro, is greater than or equal to 0, since, otherwise, `parseStream()`
+ * would have returned before the initialization of the `size` variable. Internally, this is
+ * implemented using GCC's statement expressions:
+ *
+ * https://gcc.gnu.org/onlinedocs/gcc-3.2.3/gcc/Statement-Exprs.html
+ *
+ * It is important to note that a Device OS function should never return a negative result code
+ * from a third party library as is, otherwise, it will be impossible to decode that error in the
+ * dependent code. The following code is incorrect:
+ * ```
+ * int receiveMessage(int socket) {
+ *     char buffer[128];
+ *     // It is an error to use non-Device OS functions with CHECK()
+ *     size_t size = CHECK(lwip_read(socket, buffer, sizeof(buffer)));
+ *     LOG(TRACE, "Read %u bytes", (unsigned)size);
+ *     ...
+ *     return 0;
+ * }
+ * ```
+ * This header file also defines the `CHECK_TRUE()` and `CHECK_FALSE()` macros, which can be used
```

Comment:
Please add the following phrase to begin this section.
> "For checking non-Device OS functions, this header also defines..."

Otherwise it seems to contradict the note above about not using our MACROs against non-Particle code.

Reply:
That note was about the `CHECK()` macro specifically though. Also, Device OS still has a lot of functions that return `bool`, or otherwise don't follow the convention.